File: examples/languages/python/async-batch.py
"""
Async Batch Processing

Process multiple images concurrently using asyncio.
Great for high-volume workflows.

Usage:
    python async-batch.py PATTERN_ID image1.jpg image2.jpg image3.jpg
"""

import os
import sys
import asyncio
import aiohttp
from pathlib import Path
from typing import List, Dict, Any


# Configuration
IMGGO_API_KEY = os.getenv("IMGGO_API_KEY")
IMGGO_BASE_URL = os.getenv("IMGGO_BASE_URL", "https://img-go.com/api")


async def upload_image_async(
    session: aiohttp.ClientSession,
    image_path: str,
    pattern_id: str
) -> Dict[str, Any]:
    """
    Upload a single image asynchronously.

    Args:
        session: aiohttp session
        image_path: Path to image file
        pattern_id: Pattern ID to use

    Returns:
        dict: Result with job_id, image_path, and status
    """
    try:
        print(f"Uploading: {Path(image_path).name}")

        # Read file
        with open(image_path, 'rb') as f:
            file_data = f.read()

        # Guess MIME type
        import mimetypes
        mime_type, _ = mimetypes.guess_type(image_path)

        # Create form data
        data = aiohttp.FormData()
        data.add_field('image',
                      file_data,
                      filename=Path(image_path).name,
                      content_type=mime_type or 'image/jpeg')

        # Upload
        async with session.post(
            f"{IMGGO_BASE_URL}/patterns/{pattern_id}/ingest",
            data=data
        ) as response:
            response.raise_for_status()
            result = await response.json()

            job_id = result["data"]["job_id"]
            print(f"  Created job: {job_id}")

            return {
                "job_id": job_id,
                "image_path": image_path,
                "status": "uploaded"
            }

    except Exception as e:
        print(f"  Error uploading {image_path}: {e}")
        return {
            "job_id": None,
            "image_path": image_path,
            "status": "error",
            "error": str(e)
        }


async def poll_job_async(
    session: aiohttp.ClientSession,
    job_id: str,
    image_path: str,
    max_attempts: int = 60
) -> Dict[str, Any]:
    """
    Poll a job asynchronously until completion.

    Args:
        session: aiohttp session
        job_id: Job ID to poll
        image_path: Original image path (for tracking)
        max_attempts: Maximum polling attempts

    Returns:
        dict: Result with extracted data
    """
    try:
        for attempt in range(max_attempts):
            async with session.get(
                f"{IMGGO_BASE_URL}/jobs/{job_id}"
            ) as response:
                response.raise_for_status()
                data = await response.json()

                status = data["data"]["status"]

                if status == "succeeded":
                    result = data["data"].get("manifest") or data["data"].get("result")
                    print(f"  Completed: {Path(image_path).name}")

                    return {
                        "job_id": job_id,
                        "image_path": image_path,
                        "status": "completed",
                        "result": result
                    }

                elif status == "failed":
                    error = data["data"].get("error", "Unknown error")
                    print(f"  Failed: {Path(image_path).name} - {error}")

                    return {
                        "job_id": job_id,
                        "image_path": image_path,
                        "status": "failed",
                        "error": error
                    }

                # Still processing
                await asyncio.sleep(2)

        # Timeout
        print(f"  Timeout: {Path(image_path).name}")
        return {
            "job_id": job_id,
            "image_path": image_path,
            "status": "timeout"
        }

    except Exception as e:
        print(f"  Error polling {job_id}: {e}")
        return {
            "job_id": job_id,
            "image_path": image_path,
            "status": "error",
            "error": str(e)
        }


async def process_batch(
    image_paths: List[str],
    pattern_id: str
) -> List[Dict[str, Any]]:
    """
    Process multiple images concurrently.

    Args:
        image_paths: List of paths to image files
        pattern_id: Pattern ID to use

    Returns:
        list: Results for each image
    """
    if not IMGGO_API_KEY:
        raise ValueError("IMGGO_API_KEY environment variable not set")

    # Create session with auth header
    headers = {
        "Authorization": f"Bearer {IMGGO_API_KEY}"
    }

    async with aiohttp.ClientSession(headers=headers) as session:
        # Phase 1: Upload all images concurrently
        print(f"\nUploading {len(image_paths)} images...")
        upload_tasks = [
            upload_image_async(session, image_path, pattern_id)
            for image_path in image_paths
        ]
        upload_results = await asyncio.gather(*upload_tasks)

        # Filter out failed uploads
        successful_uploads = [
            r for r in upload_results
            if r["status"] == "uploaded" and r["job_id"]
        ]

        if not successful_uploads:
            print("\nNo successful uploads!")
            return upload_results

        # Phase 2: Poll all jobs concurrently
        print(f"\nProcessing {len(successful_uploads)} jobs...")
        poll_tasks = [
            poll_job_async(session, r["job_id"], r["image_path"])
            for r in successful_uploads
        ]
        final_results = await asyncio.gather(*poll_tasks)

        return final_results


def main():
    """Main entry point"""
    if len(sys.argv) < 3:
        print("Usage: python async-batch.py <pattern_id> <image1> <image2> ...")
        print("\nExample:")
        print("  python async-batch.py pat_abc123 invoice1.jpg invoice2.jpg invoice3.jpg")
        sys.exit(1)

    pattern_id = sys.argv[1]
    image_paths = sys.argv[2:]

    # Validate files exist
    missing_files = [p for p in image_paths if not os.path.exists(p)]
    if missing_files:
        print(f"Error: Files not found: {', '.join(missing_files)}")
        sys.exit(1)

    print(f"Processing {len(image_paths)} images with pattern {pattern_id}")

    try:
        # Run async batch processing
        results = asyncio.run(process_batch(image_paths, pattern_id))

        # Display summary
        print("\n" + "="*60)
        print("BATCH PROCESSING SUMMARY")
        print("="*60)

        completed = [r for r in results if r["status"] == "completed"]
        failed = [r for r in results if r["status"] == "failed"]
        errors = [r for r in results if r["status"] == "error"]
        timeouts = [r for r in results if r["status"] == "timeout"]

        print(f"Total: {len(results)}")
        print(f"Completed: {len(completed)}")
        print(f"Failed: {len(failed)}")
        print(f"Errors: {len(errors)}")
        print(f"Timeouts: {len(timeouts)}")

        # Display failed/error details
        if failed or errors or timeouts:
            print("\nIssues:")
            for r in failed + errors + timeouts:
                error_msg = r.get("error", "Unknown")
                print(f"  - {Path(r['image_path']).name}: {r['status']} - {error_msg}")

        print("\nResults saved to: batch_results.json")

        # Save results to file
        import json
        with open("batch_results.json", "w") as f:
            json.dump(results, f, indent=2)

    except Exception as e:
        print(f"\nError: {e}")
        import traceback
        traceback.print_exc()
        sys.exit(1)


if __name__ == "__main__":
    main()

File: examples/languages/nodejs/async-batch.ts
/**
 * Async Batch Processing Example
 *
 * Process multiple images concurrently using Promise.all() for high-performance batch workflows.
 *
 * Usage:
 *   npx ts-node async-batch.ts PATTERN_ID image1.jpg image2.jpg image3.jpg ...
 *
 * Example:
 *   npx ts-node async-batch.ts pat_abc123 invoice*.jpg
 */

import axios, { AxiosError } from 'axios';
const FormData = require('form-data');
import * as fs from 'fs';
import * as path from 'path';

const IMGGO_API_KEY = process.env.IMGGO_API_KEY;
const IMGGO_BASE_URL = process.env.IMGGO_BASE_URL || 'https://img-go.com/api';
const MAX_CONCURRENT = 10; // Maximum concurrent uploads

interface JobResponse {
  data: {
    job_id: string;
  };
}

interface JobResult {
  data: {
    status: string;
    manifest?: any;
    result?: any;
    error?: string;
  };
}

interface UploadResult {
  imagePath: string;
  jobId?: string;
  error?: string;
}

interface ProcessResult {
  imagePath: string;
  success: boolean;
  data?: any;
  error?: string;
}

async function sleep(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms));
}

async function uploadImage(imagePath: string, patternId: string): Promise<UploadResult> {
  try {
    if (!fs.existsSync(imagePath)) {
      return { imagePath, error: 'File not found' };
    }

    const formData = new FormData();
    formData.append('image', fs.createReadStream(imagePath));

    const response = await axios.post<JobResponse>(
      `${IMGGO_BASE_URL}/patterns/${patternId}/ingest`,
      formData,
      {
        headers: {
          'Authorization': `Bearer ${IMGGO_API_KEY}`,
          ...formData.getHeaders(),
        },
        timeout: 30000,
      }
    );

    const jobId = response.data.data.job_id;
    console.log(`[UPLOAD] ${path.basename(imagePath)} -> ${jobId}`);
    return { imagePath, jobId };

  } catch (error) {
    const axiosError = error as AxiosError;
    const errorMessage = (axiosError.response?.data as any)?.error || axiosError.message;
    console.error(`[UPLOAD FAILED] ${path.basename(imagePath)}: ${errorMessage}`);
    return { imagePath, error: errorMessage };
  }
}

async function pollJob(
  jobId: string,
  imagePath: string,
  maxAttempts: number = 60
): Promise<ProcessResult> {
  try {
    for (let attempt = 0; attempt < maxAttempts; attempt++) {
      const response = await axios.get<JobResult>(
        `${IMGGO_BASE_URL}/jobs/${jobId}`,
        {
          headers: { 'Authorization': `Bearer ${IMGGO_API_KEY}` },
          timeout: 10000,
        }
      );

      const { status, manifest, result, error } = response.data.data;

      if (status === 'succeeded') {
        console.log(`[SUCCESS] ${path.basename(imagePath)}`);
        return {
          imagePath,
          success: true,
          data: manifest || result,
        };
      }

      if (status === 'failed') {
        console.error(`[FAILED] ${path.basename(imagePath)}: ${error}`);
        return {
          imagePath,
          success: false,
          error: error || 'Job failed',
        };
      }

      // Still processing
      if (attempt % 10 === 0) {
        console.log(`[POLLING] ${path.basename(imagePath)}: ${status} (${attempt + 1}/${maxAttempts})`);
      }

      await sleep(2000);
    }

    console.error(`[TIMEOUT] ${path.basename(imagePath)}`);
    return {
      imagePath,
      success: false,
      error: 'Job timeout',
    };

  } catch (error) {
    const axiosError = error as AxiosError;
    const errorMessage = axiosError.message;
    console.error(`[POLLING ERROR] ${path.basename(imagePath)}: ${errorMessage}`);
    return {
      imagePath,
      success: false,
      error: errorMessage,
    };
  }
}

async function processBatch(
  imagePaths: string[],
  patternId: string
): Promise<ProcessResult[]> {
  if (!IMGGO_API_KEY) {
    throw new Error('IMGGO_API_KEY environment variable not set');
  }

  console.log(`\nProcessing ${imagePaths.length} images with pattern ${patternId}...\n`);

  // Phase 1: Upload all images concurrently (in chunks)
  console.log('=== PHASE 1: UPLOADING ===\n');
  const uploadResults: UploadResult[] = [];

  for (let i = 0; i < imagePaths.length; i += MAX_CONCURRENT) {
    const chunk = imagePaths.slice(i, i + MAX_CONCURRENT);
    const chunkResults = await Promise.all(
      chunk.map(imagePath => uploadImage(imagePath, patternId))
    );
    uploadResults.push(...chunkResults);
  }

  // Separate successful uploads from failures
  const successfulUploads = uploadResults.filter(r => r.jobId);
  const failedUploads = uploadResults.filter(r => r.error);

  console.log(`\nUploaded: ${successfulUploads.length} succeeded, ${failedUploads.length} failed\n`);

  // Phase 2: Poll all jobs concurrently
  console.log('=== PHASE 2: PROCESSING ===\n');

  const pollPromises = successfulUploads.map(upload =>
    pollJob(upload.jobId!, upload.imagePath)
  );

  const processResults = await Promise.all(pollPromises);

  // Add failed uploads to results
  const allResults = [
    ...processResults,
    ...failedUploads.map(upload => ({
      imagePath: upload.imagePath,
      success: false,
      error: upload.error,
    })),
  ];

  return allResults;
}

async function main() {
  const [patternId, ...imagePaths] = process.argv.slice(2);

  if (!patternId || imagePaths.length === 0) {
    console.log('Usage: npx ts-node async-batch.ts <pattern_id> <image1> <image2> ...');
    console.log('\nExample:');
    console.log('  npx ts-node async-batch.ts pat_abc123 invoice1.jpg invoice2.jpg invoice3.jpg');
    process.exit(1);
  }

  try {
    const startTime = Date.now();
    const results = await processBatch(imagePaths, patternId);
    const endTime = Date.now();
    const duration = ((endTime - startTime) / 1000).toFixed(2);

    // Save results to JSON file
    const outputFile = `batch-results-${Date.now()}.json`;
    fs.writeFileSync(outputFile, JSON.stringify(results, null, 2));

    // Print summary
    console.log('\n' + '='.repeat(60));
    console.log('BATCH PROCESSING SUMMARY');
    console.log('='.repeat(60));

    const successful = results.filter(r => r.success);
    const failed = results.filter(r => !r.success);

    console.log(`Total Images: ${results.length}`);
    console.log(`Succeeded: ${successful.length}`);
    console.log(`Failed: ${failed.length}`);
    console.log(`Duration: ${duration}s`);
    console.log(`Output File: ${outputFile}`);

    if (failed.length > 0) {
      console.log('\nFailed Images:');
      failed.forEach(result => {
        console.log(`  - ${path.basename(result.imagePath)}: ${result.error}`);
      });
    }

    console.log('\n' + '='.repeat(60));

  } catch (error: any) {
    console.error(`\nError: ${error.message}`);
    process.exit(1);
  }
}

if (require.main === module) {
  main();
}

export { processBatch };
